pkg/asset/machines/openstack/machinesets.go

```diff
@@ -8,69 +8,71 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 
-	machinev1 "github.com/openshift/api/machine/v1"
 	clusterapi "github.com/openshift/api/machine/v1beta1"
 	"github.com/openshift/installer/pkg/types"
 	"github.com/openshift/installer/pkg/types/openstack"
-	openstackdefaults "github.com/openshift/installer/pkg/types/openstack/defaults"
 )
 
-// MachineSets returns a list of machinesets for a machinepool.
+const maxInt32 int64 = int64(^uint32(0)) >> 1
+
+// MachineSets returns the MachineSets encoded by the given machine-pool. The
+// number of returned MachineSets, while being capped to the number of
+// replicas, depends on the variable-length fields in the machine-pool. Each
+// MachineSet generates a set of identical Machines; to encode for Machines
+// spread on, say, three availability zones, three MachineSets must be
+// produced. Note that for each variable-length field (currently: Compute
+// availability zones, Storage availability zones and Root volume types), when
+// more than one is specified, values of identical index are grouped in the
+// same MachineSet.
 func MachineSets(clusterID string, config *types.InstallConfig, pool *types.MachinePool, osImage, role, userDataSecret string, clientOpts *clientconfig.ClientOpts) ([]*clusterapi.MachineSet, error) {
 	if configPlatform := config.Platform.Name(); configPlatform != openstack.Name {
 		return nil, fmt.Errorf("non-OpenStack configuration: %q", configPlatform)
 	}
 	if poolPlatform := pool.Platform.Name(); poolPlatform != openstack.Name {
 		return nil, fmt.Errorf("non-OpenStack machine-pool: %q", poolPlatform)
 	}
+	if pool.Replicas == nil || *pool.Replicas < 1 {
```

Comment:
I believe this change broke the NFV jobs, where we set the number of worker nodes to 0.
https://prow.ci.openshift.org/?job=*4.14-e2e-openstack*nfv*

We used to generate machinesets per zone, even if the number of replicas was 0, and we no longer do that. This causes the [workerSpecs to be empty](https://github.com/openshift/installer/blob/master/pkg/tfvars/openstack/openstack.go#L75C6-L85), and consequently [workerServerGroupNames](https://github.com/openshift/installer/blob/master/pkg/tfvars/openstack/openstack.go#L171-L179).

We may want to either revert these couple of lines of code where we return early in case of no replicas, or provide a default value for `openstack_worker_server_group_names` in terraform.

Reply:
https://github.com/openshift/installer/pull/7380

Reply:
@mandre I suggest we keep https://github.com/openshift/installer/pull/7359, which is a good thing anyway. WDYT?